gapii/cc/spy.cpp

```diff
@@ -97,12 +108,34 @@ Spy::Spy()
       mObserveFrameFrequency(0),
       mObserveDrawFrequency(0),
       mFrameNumber(0) {
+  // Start by checking whether to capture the current process: compare the
+  // current process name with the "capture_proc_name" that we get from the
+  // environment. An empty "capture_proc_name" means capture any process. This
+  // is useful for games where the process initially started by AGI creates an
+  // other process where the actual game rendering happens.
   bool this_executable = true;
-  char* pn = getenv("GAPID_CAPTURE_PROCESS_NAME");
-  if (pn) {
-    auto proc_name = core::get_process_name();
-    this_executable = (proc_name == pn);
+  auto this_proc_name = core::get_process_name();
+  GAPID_INFO("this process name: %s", this_proc_name.c_str());
+
+#if TARGET_OS == GAPID_OS_ANDROID
+  // PROP_VALUE_MAX is defined in <sys/system_properties.h>
+  char capture_proc_name[PROP_VALUE_MAX];
+  __system_property_get(kCaptureProcessNameSystemProperty, capture_proc_name);
+#else
+  const char* capture_proc_name = getenv(kCaptureProcessNameEnvVar);
+  // Make sure capture_proc_name is not null, since the code below assumes so.
+  if (!capture_proc_name) {
+    capture_proc_name = "";
   }
+#endif
+
+  if (strlen(capture_proc_name)) {
+    this_executable = (!strcmp(this_proc_name.c_str(), capture_proc_name));
```

Comment:
`this_proc_name` is a std::string, just use it to avoid strlen:
` this_executable = this_proc_name.compare(capture_proc_name) == 0;

Reply:
Agreed, unfortunately using string.compare() hits a bug in the NDK: https://github.com/android/ndk/issues/1494

I've added a comment.

Reply:
`core::get_process_name` has a bug on Android, it includes a trailing zero in the string, which increases the size of `std:string`, but not `strlen`.
Please fix `core/cc/android/process_name.cpp` with:
```
std::string get_process_name() {
  std::ifstream t("/proc/self/cmdline");
  std::string name;
  auto it = std::istreambuf_iterator<char>(t);
  while (it != std::istreambuf_iterator<char>() && *it != '\0') {
    name += *it++;
  }
  return name;
}
```

Reply:
And close the NDK bug and remove the comment

Reply:
Done